ledger/accountdb.go

```diff
@@ -844,34 +858,79 @@ func accountsPutTotals(tx *sql.Tx, totals AccountTotals, catchpointStaging bool)
-func accountsNewRound(tx *sql.Tx, updates map[basics.Address]accountDeltaCount, creatables map[basics.CreatableIndex]modifiedCreatable, proto config.ConsensusParams) (err error) {
+// accountsNewRound updates the accountbase and assetcreators tables by applying the provided deltas to the accounts / creatables.
+// The updates.old entries are being updated to reflect the latest value that was written to the database.
+func accountsNewRound(tx *sql.Tx, updates map[basics.Address]accountDeltaCount, creatables map[basics.CreatableIndex]modifiedCreatable, proto config.ConsensusParams, lastUpdateRound basics.Round) (err error) {
 
-	var insertCreatableIdxStmt, deleteCreatableIdxStmt, deleteStmt, replaceStmt *sql.Stmt
+	var insertCreatableIdxStmt, deleteCreatableIdxStmt, deleteByRowIDStmt, insertStmt, updateStmt *sql.Stmt
 
-	deleteStmt, err = tx.Prepare("DELETE FROM accountbase WHERE address=?")
+	deleteByRowIDStmt, err = tx.Prepare("DELETE FROM accountbase WHERE rowid=?")
 	if err != nil {
 		return
 	}
-	defer deleteStmt.Close()
+	defer deleteByRowIDStmt.Close()
 
-	replaceStmt, err = tx.Prepare("REPLACE INTO accountbase (address, normalizedonlinebalance, data) VALUES (?, ?, ?)")
+	insertStmt, err = tx.Prepare("INSERT INTO accountbase (address, normalizedonlinebalance, data) VALUES (?, ?, ?)")
 	if err != nil {
 		return
 	}
-	defer replaceStmt.Close()
+	defer insertStmt.Close()
 
+	updateStmt, err = tx.Prepare("UPDATE accountbase SET normalizedonlinebalance = ?, data = ? WHERE rowid = ?")
+	if err != nil {
+		return
+	}
+	defer updateStmt.Close()
+	var result sql.Result
+	var rowsAffected int64
 	for addr, data := range updates {
-		if data.new.IsZero() {
-			// prune empty accounts
-			_, err = deleteStmt.Exec(addr[:])
+		if data.old.rowid == 0 {
+			// zero rowid means we don't have a previous value.
+			if data.new.IsZero() {
+				// if we didn't had it before, and we don't have anything now, just skip it.
+			} else {
+				// create a new entry.
+				normBalance := data.new.NormalizedOnlineBalance(proto)
+				result, err = insertStmt.Exec(addr[:], normBalance, protocol.Encode(&data.new))
+				if err == nil {
+					data.old.rowid, err = result.LastInsertId()
+					data.old.accountData = data.new
+				}
+			}
 		} else {
-			normBalance := data.new.NormalizedOnlineBalance(proto)
-			_, err = replaceStmt.Exec(addr[:], normBalance, protocol.Encode(&data.new))
+			// non-zero rowid means we had a previous value.
+			if data.new.IsZero() {
+				// new value is zero, which means we need to delete the current value.
+				result, err = deleteByRowIDStmt.Exec(data.old.rowid)
+				if err == nil {
+					// we deleted the entry successfully.
+					data.old.rowid = 0
+					data.old.accountData = basics.AccountData{}
+					rowsAffected, err = result.RowsAffected()
+					if rowsAffected != 1 {
+						err = fmt.Errorf("failed to delete accountbase row for account %v, rowid %d", addr, data.old.rowid)
```

Comment:
mb something like "deleted more than 1 row" ?

Reply:
when making a sql statement to delete an entry by a unique primary key, there can be at most 1 row affected.. so the other option was 0. But I don't like to write `x==0` and try to pick a more extensive predicate `x != 1`